sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+   * Unlocked with R=0, W=0 and X=0.
+   *
+   * Note: full access (R=1, W=1 and X=1) in Machine Mode. Only use
+   * in disabled (OFF) entries.
+   */
+  kEpmpPermUnlocked,
+
+  /**
+   * Locked with R=0, W=0 and X=0.
+   */
+  kEpmpPermLockedNoAccess,
+
+  /**
+   * Locked with R=0, W=0 and X=1.
+   */
+  kEpmpPermLockedExecuteOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=0.
+   */
+  kEpmpPermLockedReadOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=1.
+   */
+  kEpmpPermLockedReadExecute,
+
+  /**
+   * Locked with R=1, W=1 and X=0.
+   */
+  kEpmpPermLockedReadWrite,
+} epmp_perm_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
+ * addresses).
+ *
+ * The `start` address is inclusive and the `end` address is exclusive.
+ */
+typedef struct epmp_region {
+  uintptr_t start;
+  uintptr_t end;
+} epmp_region_t;
+
+/**
+ * EPMP entry index.
+ */
+typedef enum {
+  kEpmpEntry0,
+  kEpmpEntry1,
+  kEpmpEntry2,
+  kEpmpEntry3,
+  kEpmpEntry4,
+  kEpmpEntry5,
+  kEpmpEntry6,
+  kEpmpEntry7,
+  kEpmpEntry8,
+  kEpmpEntry9,
+  kEpmpEntry10,
+  kEpmpEntry11,
+  kEpmpEntry12,
+  kEpmpEntry13,
+  kEpmpEntry14,
+  kEpmpEntry15,
+} epmp_entry_t;
+
+/**
+ * EPMP entry count.
+ */
+typedef enum {
+  kEpmpEntryCount0,
+  kEpmpEntryCount1,
+  kEpmpEntryCount2,
+  kEpmpEntryCount3,
+  kEpmpEntryCount4,
+  kEpmpEntryCount5,
+  kEpmpEntryCount6,
+  kEpmpEntryCount7,
+  kEpmpEntryCount8,
+  kEpmpEntryCount9,
+  kEpmpEntryCount10,
+  kEpmpEntryCount11,
+  kEpmpEntryCount12,
+  kEpmpEntryCount13,
+  kEpmpEntryCount14,
+  kEpmpEntryCount15,
+  kEpmpEntryCount16,
+} epmp_entry_count_t;
+
+/**
+ * EPMP generic status codes.
+ *
+ * These error codes can be used by any function. However if a function
+ * requires additional status codes, it must define a set of status codes to
+ * be used exclusively by such function.
+ */
+typedef enum {
+  kEpmpOk = 0,
+  kEpmpError,
+  kEpmpBadArg,
+} epmp_result_t;
+
+/**
+ * Initialize the EPMP library.
+ *
+ * The current state of the EPMP CSRs will be read and stored inside
+ * library. This function must be called before using the library. The
+ * library can only be initialized once.
+ *
+ * Note: it may be desirable to check the state is as expected. This
+ * can be done using `epmp_get_state`.
+ */
+epmp_result_t epmp_init(void);
+
+/**
+ * Check that the actual EPMP state matches what the library expects
+ * it to be.
+ */
+epmp_result_t epmp_check(void);
+
+/**
+ * Start an EPMP transaction.
+ *
+ * A new transaction will be initialized to expect the given number of
+ * entries to be reconfigured. Only one transaction may be in progress
+ * at a time. Each entry may only be reconfigured once per transaction.
+ * Call `epmp_transaction_end` to finalize the transaction and update
+ * the EPMP control registers.
+ *
+ * @param count Number of entries to be configured.
```

Comment:
Is there a limit to the number of transactions?

Reply:
Only one transaction may be in progress at a given time. There is currently no limit to the total number of transactions.

Reply:
> One thing we could maybe consider is having a bunch of pre-defined regions (e.g. all of ROM, all of RAM, all MMIO regions, stack, rodata, bss, data etc.). We could then represent each region with an enum. That would have the advantage of making the API harder to misuse but means adding all required regions to the library source code as we need them.

I think it would be hard to make a call about this at this point. I prefer if we continue with the API as defined in this PR and once we get to the integration stage we can evaluate if this refactoring would make sense.

Reply:
I see how `count` is useful to protect against extra calls to this library but maybe we should have two checksums: one for the state of the PMP and another for the transaction, i.e. all arguments passed during a transaction which implicitly includes the count (and we can remove `count` from this function). WDYT?

Reply:
My main concern with exposing checksums in the API is that I think we'll end up needing another API just to generate the checksums. I think there are ways we could use checksums if we moved to a code generation approach for PMP updates but I don't think it would work well if we want to keep everything in a self contained C library.